Use drainable and dynamic fillable porosity for water table change

vers-w opened the pull request:
## Issue addressed
Fixes #180

## Explanation
Drainable porosity/specific yield defined as the difference between `theta_s` (saturated water content) and `theta_fc` (water content at field capacity, new parameter for the `sbm` soil model) is now used to compute the falling water table of subsurface flow concepts `LateralSSF` and `UnconfinedAquifer`. For the rising water table of these concepts a dynamic fillable porosity (based on `theta_s` and the water content of unsaturated soil layers) is used. This results in a better integration of the `sbm` soil model (unsaturated zone) with these subsurface flow concepts as specific yield is typically represented by drainable porosity for these concepts and not by the effective porosity (difference between saturated and residual water content) in the current implementation.

## Checklist
- [x] Updated tests or added new tests
- [x] Branch is up to date with `master`
- [x] Tests & pre-commit hooks pass
- [x] Updated documentation if needed
- [x] Updated changelog.qmd if needed

## Additional Notes (optional)
Add any additional notes or information that may be helpful.

vers-w committed:
Add drainable porosity to `sbm` soil model

vers-w committed:
Make use of varying `specific_yield` for `LateralSSF`
For a rising water table this is based on water content (soil layers) and `theta_s`, for a falling water table it is equal to `theta_s` - `theta_d`.

vers-w committed:
Use varying specific yield for `UnconfinedAquifer` and couple of fixes
As for `LateralSSF`, for a rising water table this is based on water content (soil layers) and `theta_s`, for a falling water table it is equal to `theta_s` - `theta_d`. This results in a better integration with the `SbmSoilModel` (unsaturated zone). The update of `ustorelayerdepth` has been moved to a separate function and is called from the update of the subsurface flow model (`UnconfinedAquifer` can have a subdaily timestep). Fixed computation of dynamic specific yield for a rising water table (was computed over each unsaturated soil layer above the water table, and should be only computed for unsaturated soil layers that satisfy net flux from subsurface flow).  In addition, for a rising water table a check has been added for a saturated layer above the water table, in that case `dh` is equal to the layer thickness. The kinematic wave subsurface flow for different `kh_profile` types has been merged into one function, and for `KhExponential` and `KhExponentialConstant` the kinematic wave celerity at the start of the timestep is now used (simplification). Update of  tests.

vers-w committed:
Merge branch 'master' into feat/drainable_porosity

vers-w committed:
Input parameter drainable porosity
Drainable porosity can be read from input, else it is computed.

vers-w committed:
Rename drainable porosity to field capacity
Field capacity is used to compute drainable porosity/specific yield.

vers-w committed:
Use field capacity parameter for `NonPaddy`
Remove computation of field capacity as part of `NonPaddy` gross water demand computation.

vers-w committed:
Test unconfined `update!` function groundwater

vers-w committed:
Merge branch 'master' into feat/drainable_porosity

vers-w committed:
Merge branch 'master' into feat/drainable_porosity

vers-w committed:
This file was missing in previous commit

vers-w committed:
Add docstring and argument types
Argument types are added to functions `update_head` and `water_table_change`.

vers-w committed:
Merge branch 'master' into feat/drainable_porosity

vers-w committed:
Apply pre-commit changes

vers-w committed:
Set lower limit for drainable porosity

vers-w committed:
Add field capacity parameter to docs

vers-w committed:
Update field capacity description docs
In section non-paddy irrigation.

vers-w committed:
Update docs Land Hydrology `SBM`
Add section on water table change and input parameters to compute this.

vers-w committed:
Update docs kinwave.qmd
Subsurface flow routing.

vers-w committed:
Merge branch 'master' into feat/drainable_porosity

vers-w committed:
Update docs subsurface flow routing
Explain how water table depth is computed.

vers-w committed:
Update doc strings

vers-w committed:
Update docs gwf.qmd

vers-w committed:
Set lower limit for drainable porosity model type `sbm_gwf`

vers-w committed:
Remove specific yield input parameter
For the model type `sbm_gwf` (unconfined aquifer) specific yield is based on soil `SBM` parameters `theta_s` and `theta_fc`. This is used for a falling water table, for a rising water table fillable porosity is computed.

vers-w committed:
Merge branch 'master' into feat/drainable_porosity

vers-w committed:
Missing files merging master branch

vers-w committed:
Fix docstrings

vers-w committed:
Merge branch 'master' into feat/drainable_porosity

vers-w committed:
Fix type instability `boundaries` of `GroundwaterFlow`

vers-w committed:
Let's also commit test files...

vers-w committed:
Rename `GroundwaterFlowBC` to `AquiferBC`

vers-w committed:
Update changelog

vers-w committed:
Merge branch 'master' into feat/drainable_porosity

SouthEndMusic commented on Wflow/src/mass_balance.jl:
```diff
@@ -8,10 +8,22 @@ time step `storage_prev` required for mass balance computation.
 @with_kw struct MassBalance <: AbstractMassBalance
     n::Int # number of cells/nodes
     storage_prev::Vector{Float64} = zeros(n)
+    zi_prev::Vector{Float64} = []
```
Are these trivial defaults?

JoostBuitink commented on Wflow/src/groundwater/aquifer.jl:
```diff
@@ -121,12 +121,13 @@ instead. Specific yield will vary roughly between 0.05 (clay) and 0.45 (peat)
+    bottom::Vector{Float64}             # bottom of groundwater layer [m]
+    area::Vector{Float64}               # area of cell [m²]
+    specific_yield::Vector{Float64}     # [m m⁻¹]
+    specific_yield_dyn::Vector{Float64} # [m m⁻¹]
```
Maybe good to add descriptions here as well

vers-w committed:
Merge branch 'master' into feat/drainable_porosity

vers-w committed:
Fix limiting subsurface flow
Fixed limiting kinematic wave subsurface flow partly below the soil bottom.

vers-w committed:
Fix unit conversion `recharge`
As part of `MassBalance` computations for `LateralSSF`.

vers-w committed:
Fix check unsaturated layer fully saturated
To check if an unsaturated soil layer is fully saturated with dynamic specific yield `sy` (`sy` = 0) did not work because of floating point errors, as part of function `water_table_change`. This also results in water balance issues. Replacing the check with `flux_layer` (subsurface water flux to soil layer) fixes this issue.

vers-w committed:
Introduce variable `drainable_waterdepth`
To keep track of the amount of water that can drain from the saturated zone.

vers-w committed:
Use function for lower bound drainable porosity

vers-w committed:
Saturated store and `drainable_waterdepth`
Use `drainable_waterdepth` to limit fluxes from saturated store.

vers-w committed:
Improve check unsaturated layer fully saturated

vers-w committed:
Update docs

vers-w committed:
Merge branch 'master' into feat/drainable_porosity

vers-w committed:
Add type restriction for `soil` function argument
For function ` kinematic_wave_ssf`.

vers-w committed:
Simplify function `update_head`
Use `soil` as function argument.

vers-w committed:
Also commit this file...

vers-w committed:
Simplify function `water_table_change`
Use `soil` as function argument.

vers-w committed:
Fix effective porosity kinematic wave ssf
Should be based on `theta_s` - `theta_fc` and not `theta_s` - `theta_r`.

vers-w committed:
Also fix effective porosity
For `KhExponential` and `KhExponentialConstant` profiles.

JoostBuitink reviewed (approved):
Only two small comments! All changes look good to me, with quite some effects on the resulting discharge (as discussed).

vers-w committed:
Address review comment

vers-w committed:
Add release note to changelog

AquaPore commented:
Dear @vers-w and @JoostBuitink , 

I want to thank you for making this improvement, and indeed I agree that _specific yield_ in the groundwater module is more physical when :

$SpecificYield= \theta (h=0) - \theta(h=h  \_{fc} )$ . 

Nevertheless, I am a bit concerned that when we introduce a new parameter $\theta _{fc}$  it can be **disconnected** to the derived $\theta(h)$ relationship of Brooks and Corey.

Instead, to remain "physical" I propose to remplace  $\theta _{fc}$ with $h  \_{fc}$ as a parameter, as it is well accepted that the tension at which field capacity occurs varies between -10kPa and -33kPa depending if the soil is coarse or fine.

Kind regards,
Joseph A.P. Pollacco
